File: integrations/src/main/java/services/voice/service/CallLogService.java
package services.voice.service;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import services.voice.dto.CallLogRequest;
import services.voice.entity.CallLog;
import services.voice.repository.CallLogRepository;

import java.time.LocalDateTime;
import java.util.List;

@Service
@Transactional
public class CallLogService {

    private final CallLogRepository callLogRepository;

    @Autowired
    public CallLogService(CallLogRepository callLogRepository) {
        this.callLogRepository = callLogRepository;
    }

    public CallLog saveCallLog(CallLog callLog) {
        return callLogRepository.save(callLog);
    }

    public CallLog updateCallStatus(String callSid, String status, String recordingUrl) {
        return callLogRepository.findByCallSid(callSid)
                .map(callLog -> {
                    callLog.setStatus(status);
                    if (recordingUrl != null) {
                        callLog.setRecordingUrl(recordingUrl);
                    }
                    return callLogRepository.save(callLog);
                })
                .orElse(null);
    }

    public CallLog updateCallStatus(String callSid, String status) {
        return updateCallStatus(callSid, status, null);
    }

    public CallLog updateCallEndTime(String callSid, LocalDateTime endTime, Integer duration) {
        return callLogRepository.findByCallSid(callSid)
                .map(callLog -> {
                    callLog.setEndTime(endTime);
                    if (duration != null) {
                        callLog.setDuration(duration);
                    }
                    return callLogRepository.save(callLog);
                })
                .orElse(null);
    }

    public CallLog updateCallEndTime(String callSid, LocalDateTime endTime) {
        return updateCallEndTime(callSid, endTime, null);
    }

    public CallLog getCallLogBySid(String callSid) {
        return callLogRepository.findByCallSid(callSid).orElse(null);
    }

    public List<CallLog> getCallLogsByEntity(String entityName, Long entityId) {
        return callLogRepository.findByEntity(entityName, entityId);
    }

    public List<CallLog> getCallLogsByPhoneNumber(String phoneNumber) {
        return callLogRepository.findByPhoneNumber(phoneNumber, phoneNumber);
    }

    public List<CallLog> getCallLogsByStatus(String status) {
        return callLogRepository.findByStatus(status);
    }

    public List<CallLog> getCallLogsByDirection(String direction) {
        return callLogRepository.findByDirection(direction);
    }

    public CallLog createCallLog(CallLogRequest request) {
        CallLog callLog = new CallLog();
        callLog.setCallSid(request.getCallSid());
        callLog.setEntityName(request.getEntityName());
        callLog.setEntityId(request.getEntityId());
        callLog.setDirection(request.getDirection());
        callLog.setFromNumber(request.getFromNumber());
        callLog.setToNumber(request.getToNumber());
        callLog.setStatus(request.getStatus());
        callLog.setExophone(request.getExophone());
        callLog.setStartTime(LocalDateTime.now());
        return saveCallLog(callLog);
    }
}

File: integrations/src/main/kotlin/services/voice/service/CallLogService.kt
package services.voice.service

import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import services.voice.dto.CallLogRequest
import services.voice.entity.CallLog
import services.voice.repository.CallLogRepository
import java.time.LocalDateTime

@Service
@Transactional
class CallLogService(
    private val callLogRepository: CallLogRepository
) {

    fun saveCallLog(callLog: CallLog): CallLog {
        return callLogRepository.save(callLog)
    }

    fun updateCallStatus(callSid: String, status: String, recordingUrl: String? = null): CallLog? {
        val callLog = callLogRepository.findByCallSid(callSid)
        return if (callLog != null) {
            callLog.status = status
            if (recordingUrl != null) {
                callLog.recordingUrl = recordingUrl
            }
            callLogRepository.save(callLog)
        } else {
            null
        }
    }

    fun updateCallEndTime(callSid: String, endTime: LocalDateTime, duration: Int? = null): CallLog? {
        val callLog = callLogRepository.findByCallSid(callSid)
        return if (callLog != null) {
            callLog.endTime = endTime
            if (duration != null) {
                callLog.duration = duration
            }
            callLogRepository.save(callLog)
        } else {
            null
        }
    }

    fun getCallLogBySid(callSid: String): CallLog? {
        return callLogRepository.findByCallSid(callSid)
    }

    fun getCallLogsByEntity(entityName: String, entityId: Long): List<CallLog> {
        return callLogRepository.findByEntity(entityName, entityId)
    }

    fun getCallLogsByPhoneNumber(phoneNumber: String): List<CallLog> {
        return callLogRepository.findByPhoneNumber(phoneNumber, phoneNumber)
    }

    fun getCallLogsByStatus(status: String): List<CallLog> {
        return callLogRepository.findByStatus(status)
    }

    fun getCallLogsByDirection(direction: String): List<CallLog> {
        return callLogRepository.findByDirection(direction)
    }

    fun createCallLog(request: CallLogRequest): CallLog {
        val callLog = CallLog(
            callSid = request.callSid,
            entityName = request.entityName,
            entityId = request.entityId,
            direction = request.direction,
            fromNumber = request.fromNumber,
            toNumber = request.toNumber,
            status = request.status,
            exophone = request.exophone,
            startTime = LocalDateTime.now()
        )
        return saveCallLog(callLog)
    }
}
